File: tgboost/gbm.py
import numpy as np
from loss import SquareLoss, LogisticLoss, CustomizeLoss
from tree import Tree
from metric import get_metric
from attribute_list import AttributeList
from class_list import ClassList
from binning import BinStructure
from sampling import RowSampler, ColumnSampler
import logging


logging.basicConfig(level=logging.INFO, format='%(asctime)s : %(message)s', datefmt="[%H:%M:%S]")


class TGBoost(object):
    """
    Tiny Gradient Boosting
    """
    def __init__(self):
        self.trees = []
        self.eta = None
        self.num_boost_round = None
        self.first_round_pred = None
        self.loss = None
        self.max_depth = None
        self.subsample = None
        self.colsample = None
        self.reg_lambda = None
        self.min_sample_split = None
        self.gamma = None
        self.num_thread = None
        self.min_child_weight = None
        self.scale_pos_weight = None
        self.eval_metric = None

    def fit(self,
            features,
            label,
            validation_data=(None, None),
            early_stopping_rounds=np.inf,
            maximize=True,
            eval_metric=None,
            loss="logisticloss",
            eta=0.3,
            num_boost_round=1000,
            max_depth=6,
            scale_pos_weight=1,
            subsample=0.8,
            colsample=0.8,
            min_child_weight=1,
            min_sample_split=10,
            reg_lambda=1.0,
            gamma=0,
            num_thread=-1):

        """
        :param features: np.array
        :param label: np.array
        :param eta: learning rate
        :param num_boost_round: number of boosting round
        :param max_depth: max depth of each tree
        :param subsample: row sample rate when building a tree
        :param colsample: column sample rate when building a tree
        :param min_sample_split: min number of samples in a leaf node
        :param loss: loss object
                     logisticloss,squareloss, or customize loss
        :param reg_lambda: lambda
        :param gamma: gamma
        :param num_thread: number of threself.tree_predict_Xad to parallel
        :param eval_metric: evaluation metric, provided: "accuracy"
        """
        self.eta = eta
        self.num_boost_round = num_boost_round
        self.max_depth = max_depth
        self.subsample = subsample
        self.colsample = colsample
        self.reg_lambda = reg_lambda
        self.gamma = gamma
        self.min_sample_split = min_sample_split
        self.num_thread = num_thread
        self.eval_metric = eval_metric
        self.min_child_weight = min_child_weight
        self.scale_pos_weight = scale_pos_weight
        self.first_round_pred = 0

        # initial loss function
        if loss == "logisticloss":
            self.loss = LogisticLoss()
        elif loss == "squareloss":
            self.loss = SquareLoss()
            self.first_round_pred = label.mean()
        else:
            try:
                self.loss = CustomizeLoss(loss)
            except:
                raise NotImplementedError("loss should be 'logisticloss','squareloss', or customize loss function")

        # initialize row_sampler, col_sampler, bin_structure, attribute_list, class_list
        row_sampler = RowSampler(features.shape[0], self.subsample)
        col_sampler = ColumnSampler(features.shape[1], self.colsample)
        bin_structure = BinStructure(features)
        attribute_list = AttributeList(features, bin_structure)
        class_list = ClassList(label)
        class_list.initialize_pred(self.first_round_pred)
        class_list.update_grad_hess(self.loss)

        # to evaluate on validation set and conduct early stopping
        # we should get (val_features,val_label)
        # and set some variable to check when to stop
        do_validation = True
        if not isinstance(validation_data, tuple):
            raise TypeError("validation_data should be (val_features, val_label)")

        val_features, val_label = validation_data
        val_pred = None
        if val_features is None or val_label is None:
            do_validation = False
        else:
            val_pred = np.ones(val_label.shape) * self.first_round_pred

        if maximize:
            best_val_metric = - np.inf
            best_round = 0
            become_worse_round = 0
        else:
            best_val_metric = np.inf
            best_round = 0
            become_worse_round = 0

        # start learning
        logging.info("TGBoost start training")
        for i in range(self.num_boost_round):
            # train current tree
            tree = Tree(self.min_sample_split,
                        self.min_child_weight,
                        self.max_depth,
                        self.colsample,
                        self.subsample,
                        self.reg_lambda,
                        self.gamma,
                        self.num_thread)
            tree.fit(attribute_list, class_list, row_sampler, col_sampler, bin_structure)

            # when finish building this tree, update the class_list.pred, grad, hess
            class_list.update_pred(self.eta)
            class_list.update_grad_hess(self.loss)

            # save this tree
            self.trees.append(tree)
            logging.debug("current tree has {} nodes, {} nan tree nodes".format(tree.nodes_cnt, tree.nan_nodes_cnt))

            # print training information
            if self.eval_metric is None:
                logging.info("TGBoost round {iteration}".format(iteration=i))
            else:
                try:
                    mertric_func = get_metric(self.eval_metric)
                except:
                    raise NotImplementedError("The given eval_metric is not provided")

                train_metric = mertric_func(self.loss.transform(class_list.pred), label)

                if not do_validation:
                    logging.info("TGBoost round {iteration}, train-{eval_metric}: {train_metric:.4f}".format(
                        iteration=i, eval_metric=self.eval_metric, train_metric=train_metric))
                else:
                    val_pred += self.eta * tree.predict(val_features)
                    val_metric = mertric_func(self.loss.transform(val_pred), val_label)
                    logging.info("TGBoost round {iteration}, train-{eval_metric}: {train_metric:.4f}, val-{eval_metric}: {val_metric:.4f}".format(
                        iteration=i, eval_metric=self.eval_metric, train_metric=train_metric, val_metric=val_metric))

                    # check whether to early stop
                    if maximize:
                        if val_metric > best_val_metric:
                            best_val_metric = val_metric
                            best_round = i
                            become_worse_round = 0
                        else:
                            become_worse_round += 1
                        if become_worse_round > early_stopping_rounds:
                            logging.info("TGBoost training Stop, best round is {best_round}, best {eval_metric} is {best_val_metric:.4f}".format(
                                best_round=best_round, eval_metric=eval_metric, best_val_metric=best_val_metric))
                            break
                    else:
                        if val_metric < best_val_metric:
                            best_val_metric = val_metric
                            best_round = i
                            become_worse_round = 0
                        else:
                            become_worse_round += 1
                        if become_worse_round > early_stopping_rounds:
                            logging.info("TGBoost training Stop, best round is {best_round}, best val-{eval_metric} is {best_val_metric:.4f}".format(
                                best_round=best_round, eval_metric=eval_metric, best_val_metric=best_val_metric))
                            break

    def predict(self, features):
        assert len(self.trees) > 0
        preds = np.zeros((features.shape[0],))
        preds += self.first_round_pred
        for tree in self.trees:
            preds += self.eta * tree.predict(features)
        return self.loss.transform(preds)

File: src/main/GBM.java
package main;

import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.logging.Level;
import java.util.logging.Logger;


public class GBM {
    private ArrayList<Tree> trees = new ArrayList<>();
    private double eta;
    private int num_boost_round;
    private double first_round_pred;
    private Loss loss;
    private int max_depth;
    private double rowsample;
    private double colsample;
    private double lambda;
    private int min_sample_split;
    private double gamma;
    private int num_thread;
    private double min_child_weight;
    private double scale_pos_weight;
    private String eval_metric;

    private static Logger logger = Logger.getLogger("InfoLogging");

    public GBM(){};
    public GBM(ArrayList<Tree> trees, Loss loss, double first_round_pred, double eta){
        this.trees = trees;
        this.loss = loss;
        this.first_round_pred = first_round_pred;
        this.eta = eta;
    }

    public void fit(String file_training,
                    String file_validation,
                    ArrayList<String> categorical_features,
                    int early_stopping_rounds,
                    boolean maximize,
                    String eval_metric,
                    String loss,
                    double eta,
                    int num_boost_round,
                    int max_depth,
                    double scale_pos_weight,
                    double rowsample,
                    double colsample,
                    double min_child_weight,
                    int min_sample_split,
                    double lambda,
                    double gamma,
                    int num_thread){
        this.eta = eta;
        this.num_boost_round = num_boost_round;
        this.max_depth = max_depth;
        this.rowsample = rowsample;
        this.colsample = colsample;
        this.lambda = lambda;
        this.gamma = gamma;
        this.min_sample_split = min_sample_split;
        this.num_thread = num_thread;
        this.eval_metric = eval_metric;
        this.min_child_weight = min_child_weight;
        this.scale_pos_weight = scale_pos_weight;

        TrainData trainset = new TrainData(file_training,categorical_features);
        AttributeList attribute_list = new AttributeList(trainset);
        ClassList class_list = new ClassList(trainset);
        RowSampler row_sampler = new RowSampler(trainset.dataset_size,this.rowsample);
        ColumnSampler col_sampler = new ColumnSampler(trainset.feature_dim,this.colsample);
        trainset = null;

        if(loss.equals("logloss")){
            this.loss = new LogisticLoss();
            this.first_round_pred = 0.0;
        }else if(loss.equals("squareloss")){
            this.loss = new SquareLoss();
            this.first_round_pred = average(class_list.label);
        }
        class_list.initialize_pred(this.first_round_pred);

        class_list.update_grad_hess(this.loss,this.scale_pos_weight);

        //to evaluate on validation set and conduct early stopping
        boolean do_validation;
        ValidationData valset;
        double[] val_pred;
        if(file_validation.equals("")){
            do_validation = false;
            valset = null;
            val_pred = null;
        }else {
            do_validation = true;
            valset = new ValidationData(file_validation);
            val_pred = new double[valset.dataset_size];
            Arrays.fill(val_pred,this.first_round_pred);
        }

        double best_val_metric;
        int best_round=0;
        int become_worse_round=0;
        if(maximize){
            best_val_metric = -Double.MAX_VALUE;
        }else {
            best_val_metric = Double.MAX_VALUE;
        }

        //Start learning
        logger.info("TGBoost start training");
        for(int i=0;i<num_boost_round;i++){
            Tree tree = new Tree(min_sample_split,min_child_weight,max_depth,colsample,rowsample,
                                 lambda,gamma,num_thread,attribute_list.cat_features_cols);
            tree.fit(attribute_list,class_list,row_sampler,col_sampler);
            //when finish building this tree, update the class_list.pred, grad, hess
            class_list.update_pred(this.eta);
            class_list.update_grad_hess(this.loss, this.scale_pos_weight);


            //save this tree
            this.trees.add(tree);

            logger.log(Level.INFO,
                    String.format("current tree has %d nodes,including %d nan tree nodes",tree.nodes_cnt,tree.nan_nodes_cnt));

            //print training information
            if(eval_metric.equals("")){
                logger.log(Level.FINEST,String.format("TGBoost round %d",i));
            }else {
                double train_metric = calculate_metric(eval_metric,this.loss.transform(class_list.pred),class_list.label);


                if(!do_validation){
                    logger.log(Level.INFO,String.format("TGBoost round %d,train-%s:%.6f",i,eval_metric,train_metric));
                }else {
                    double[] cur_tree_pred = tree.predict(valset.origin_feature);
                    for(int n=0;n<val_pred.length;n++){
                        val_pred[n] += this.eta * cur_tree_pred[n];
                    }
                    double val_metric = calculate_metric(eval_metric,this.loss.transform(val_pred),valset.label);
                    logger.log(Level.INFO,String.format("TGBoost round %d,train-%s:%.6f,val-%s:%.6f",i,eval_metric,train_metric,eval_metric,val_metric));
                    //check whether to early stop
                    if(maximize){
                        if(val_metric>best_val_metric){
                            best_val_metric = val_metric;
                            best_round = i;
                            become_worse_round = 0;
                        }else {
                            become_worse_round += 1;
                        }
                        if(become_worse_round>early_stopping_rounds){
                            logger.log(Level.INFO,String.format("TGBoost training stop,best round is %d,best val-%s is %.6f",i,eval_metric,best_val_metric));
                            break;
                        }
                    }else{
                        if(val_metric<best_val_metric){
                            best_val_metric = val_metric;
                            best_round = i;
                            become_worse_round = 0;
                        }else {
                            become_worse_round += 1;
                        }
                        if(become_worse_round>early_stopping_rounds){
                            logger.log(Level.INFO,String.format("TGBoost training stop,best round is %d,best val-%s is %.6f",i,eval_metric,best_val_metric));
                            break;
                        }
                    }
                }
            }
        }
    }

    public double[] predict(float[][] features){
        logger.info("TGBoost start predicting...");
        double[] pred = new double[features.length];
        for(int i=0;i<pred.length;i++){
            pred[i] += first_round_pred;
        }
        for(Tree tree:this.trees){
            double[] cur_tree_pred = tree.predict(features);
            for(int i=0;i<pred.length;i++){
                pred[i] += this.eta * cur_tree_pred[i];
            }
        }
        return this.loss.transform(pred);
    }

    public void predict(String file_test,String file_output){
        TestData testdata = new TestData(file_test);
        double[] preds = this.predict(testdata.origin_feature);
        String[] strs = new String[preds.length];
        for(int i=0;i<strs.length;i++){
            strs[i] = String.valueOf(preds[i]);
        }
        String content = String.join("\n",strs);
        try{
            Files.write(Paths.get(file_output), content.getBytes());
        }catch (IOException e){
            e.printStackTrace();
        }

    }

    private double calculate_metric(String eval_metric,double[] pred,double[] label){
        if(eval_metric.equals("acc")){
            return Metric.accuracy(pred,label);
        }else if(eval_metric.equals("error")){
            return Metric.error(pred,label);
        }else if(eval_metric.equals("mse")){
            return Metric.mean_square_error(pred,label);
        }else if(eval_metric.equals("mae")){
            return Metric.mean_absolute_error(pred,label);
        }else if(eval_metric.equals("auc")){
            return Metric.auc(pred,label);
        }else {
            throw new NotImplementedException();
        }
    }


    private double average(double[] vals){
        double sum = 0.0;
        for(double v:vals){
            sum += v;
        }
        return sum/vals.length;
    }

    public double getFirst_round_pred(){
        return this.first_round_pred;
    }

    public double getEta(){
        return this.eta;
    }

    public Loss getLoss(){
        return this.loss;
    }

    public ArrayList<Tree> getTrees() {
        return this.trees;
    }
}
